src/AutoMaxLineJob.sol

```diff
@@ -0,0 +1,103 @@
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+pragma solidity 0.8.13;
+
+import {IJob} from "./interfaces/IJob.sol";
+
+interface SequencerLike {
+    function isMaster(bytes32 network) external view returns (bool);
+}
+
+interface AutoMaxLineLike {
+    function autoLine() external returns (address);
+    function exec() external returns (
+        uint256 oldMaxLine,
+        uint256 newMaxLine,
+        uint256 debt,
+        uint256 oldDuty,
+        uint256 newDuty
+    );
+}
+
+interface AutoLineLike {
+    function ilks(bytes32) external view returns (uint256, uint256, uint48, uint48, uint48);
+}
+
+/// @title Trigger automaxline updates based on thresholds
+contract AutoMaxLineJob is IJob {
+
+    uint256 constant internal BPS = 10 ** 4;
+    
+    SequencerLike   public immutable sequencer;
+    AutoMaxLineLike public immutable automaxline;
+    AutoLineLike    public immutable autoline;
+    bytes32         public immutable ilk;
+    uint256         public immutable thi; // % above the previous max line level
+    uint256         public immutable tlo; // % below the previous max line level
+
+    // --- Errors ---
+    error NotMaster(bytes32 network);
+    error OutsideThreshold(uint256 oldMaxLine, uint256 newMaxLine, uint256 gap);
+    error OutsideDebtRange(uint256 oldMaxLine, uint256 newMaxLine, uint256 debt, uint256 gap);
+
+    // --- Events ---
+    event Work(bytes32 indexed network);
+
+    constructor(address _sequencer, address _automaxline, bytes32 _ilk, uint256 _thi, uint256 _tlo) {
+        sequencer   = SequencerLike(_sequencer);
+        automaxline = AutoMaxLineLike(_automaxline);
+        autoline    = AutoLineLike(automaxline.autoLine());
+        ilk = _ilk;
+        thi = _thi;
+        tlo = _tlo;
+    }
+
+    function work(bytes32 network, bytes calldata) external override {
+        if (!sequencer.isMaster(network)) revert NotMaster(network);
+
+        (uint256 oldMaxLine, uint256 newMaxLine, uint256 debt,,) = automaxline.exec();
+        (, uint256 gap,,,) = autoline.ilks(ilk);
+
+        // Execution is not enough
+        // We need to be over the threshold amounts
+        if (
+            newMaxLine < oldMaxLine + gap * thi / BPS &&
+            newMaxLine + gap * tlo / BPS > oldMaxLine
+        ) revert OutsideThreshold(oldMaxLine, newMaxLine, gap);
```

Comment:
Should we not check if the duty needs updating here? Otherwise if there is a small change in maxLine that falls within the threshold range but is still able to cross the `debt` level, the duty update would not be triggered.

````suggestion
        if (
            newDuty == oldDuty &&
            newMaxLine < oldMaxLine + gap * thi / BPS &&
            newMaxLine + gap * tlo / BPS > oldMaxLine
        ) revert OutsideThreshold(oldMaxLine, newMaxLine, gap);

Reply:
I mean I don't think it's wrong to ignore the duty update here but we may want to help users get outside the wind-down territory ASAP to help them save on fees and also maybe avoid situations where a bad actor could manually trigger the wind-down fees to help liquidate vaults when the debt is just under the maxLine and occasionally jumps over it as there would not be a "good" actor counteracting them by setting the fees in the other direction. Obviously the trade-off is keepers having to pay more in gas in that edge-case.

Reply:
This is worth thinking more about, but my main concern with regarding to the duty was having a kind of dancing behavior of debt around maxLine.

One example is if line is at 57.1M, max line is 57M, and debt is 56.999M.
Note that in this situation, although max line is lower than line, dss-cron current auto-line keeper job will probably not be triggered since they are very close.

In this situation probably ppl would draw more debt to reach say 57.001M, but then we trigger and duty will spike.
So ppl will payback debt to ~56.999M, and then we trigger and duty will return to normal.
Then over and over again..

That's why it felt saver just to regard the max line, which is not affected by the debt, just uniswap (and has thresholds to avoid triggering too often).

Reply:
I see what you mean but tbh ignoring the debt means that if the duty spikes and then the max line stays constant for some time (which seems very possible), users will have no other option than manually call exec() as repaying the debt will have no effect on the keeper. In fact, due to the next check (OutsideDebtRange), a whale repaying a large amount of debt risks making it even less likely for this job to be triggered as the new debt + gap might end up being lower than the max line.